Add and update hyperfunctions for Toolkit 1.6

charislam opened the pull request:
# Description

Add and update hyperfunctions for Toolkit 1.6.

# Version

Which documentation version does this PR apply to?

- [x] Latest (Default)
- [ ] Version 1.7
- [ ] Older [specify]

# Links

Fixes https://github.com/timescale/docs-private/issues/26

charislam committed:
Draft topn_agg and gauge_agg, plus changes to topn

charislam commented on api/topn_agg.md:
````diff
@@ -0,0 +1,71 @@
+calculate the `n` most-frequent values in a column.
+```sql
+topn_agg (
+    FIXME INTEGER,
````
Didn't know the name of this parameter and just guessed at the type

epgts replied:
It is an `INTEGER` and it is the `n` referred to in the paragraph above, so I would suggest calling it `n` here as well.

charislam commented on api/topn_agg.md:
```diff
@@ -0,0 +1,71 @@
+topn_agg (
+    FIXME INTEGER,
+    value AnyElement
+) RETURNS TopnAggregate
```
Guessed at the name of the return parameter by extrapolation

epgts replied:
It's called SpaceSavingAggregate. It's returned by both topn_agg and freq_agg.

charislam commented on api/topn_agg.md:
```diff
@@ -0,0 +1,71 @@
+
+|Name|Type|Description|
+|-|-|-|
+|`FIXME`|`INTEGER`|The target number of most-frequent values|
```
Does `topn_agg` only store `n` values, or does it store more, in case it needs them for re-aggregations?

epgts replied:
It may store more than n; @WireBaron ?

WireBaron replied:
It tracks significantly more than N, but we'll only return an error if you pass a value greater than `n` to `topn` with a `topn_agg`, if that makes any sense.

charislam commented:
@WireBaron, here's a draft for `topn_agg`, `gauge_agg`, and the changes to `topn`. I left a few questions where I was unsure, and I'm happy to have any corrections where things are inaccurate or oddly worded

Loquacity committed:
Merge branch 'latest' into charis/hyperfunctions-1-5

epgts commented on api/gauge_agg.md:
```diff
@@ -0,0 +1,76 @@
+<highlight type="note">
+Both `ts` and `value` can be `NULL`, but the aggregate is not evaluated on
+`NULL` values. If the aggregate receives only a `NULL` value, it returns
+`NULL`. If it receives both non-`NULL` and `NULL` values, it ignores the `NULL`
```
While this may be our goal, it doesn't seem to be true for any of our aggregates any longer due to https://github.com/timescale/timescaledb-toolkit/issues/362

I just added a test for state_agg and can confirm it is suffering the same bug.

charislam replied:
Answered my own question :) I tested with `counter_agg` and it worked if some values were `NULL` but not if all values were `NULL`

Loquacity committed:
Merge branch 'latest' into charis/hyperfunctions-1-5

charislam committed:
Correct description of behavior with all NULLs

Loquacity committed:
Merge branch 'latest' into charis/hyperfunctions-1-5

epgts commented on api/counter_agg_point.md:
```diff
@@ -19,10 +19,8 @@ as a different numeric type you can cast to DOUBLE PRECISION on input to the
-returns NULL, it does not return an error. If non-NULL values are also received, the NULL 
-values are ignored. Both `ts` and `value` must be non-NULL for the row to be included.
+If there are `NULL` values in your data, the aggregate ignores them. At least
+some of the values must be non-`NULL`. Otherwise, you get an error.
```
I don't know if it makes sense to call out that this is a bug: https://github.com/timescale/timescaledb-toolkit/issues/362

charislam replied:
Though it's a bug, the actual behavior doesn't defy all expectations, and this is in the experimental schema, so I'd just keep it simple and not mention it. When the bug is fixed, we can just change the callout.

charislam committed:
Minor changes to update argument names, etc

charislam commented on api/topn_agg.md:
```diff
@@ -0,0 +1,71 @@
+## Optional arguments
+|Name|Type|Description|
+|-|-|-|
+|`skew`|`DOUBLE PRECISION`|The estimated skew of the data, defined as the `s` parameter of a zeta distribution. Defaults to `1.1`.|
```
```suggestion
|`skew`|`DOUBLE PRECISION`|The estimated skew of the data, defined as the `s` parameter of a zeta distribution. Must be greater than `1.0`. Defaults to `1.1`.|
```

charislam committed:
Add constraint to `skew` description

WireBaron commented on api/topn.md:
```diff
@@ -14,18 +33,31 @@ do not use any experimental features in production.
 
-If fewer than `N` elements have been found with the minimum
-frequency of the aggregate, this returns fewer than `N` values.
+If a frequency aggregate is used, and it contains fewer than `n` elements, the
+function returns fewer than `n` values.
```
```suggestion
If there are fewer than `n` value available, either because there are not 
enough elements with the minimum frequency in a frequency aggregate, 
or the data isn't skewed enough to support a top n aggregate, the function 
returns fewer than `n` values.

Requesting more values from a top n aggregate than it was created for will
return an error.
```

Loquacity committed:
Merge branch 'latest' into charis/hyperfunctions-1-5

charislam commented on api/topn.md:
```diff
@@ -14,18 +33,31 @@ do not use any experimental features in production.
+Because frequency aggregates and top N aggregates can operate over data of any
+type, `topn` uses the `ty` parameter to specify the data type of the output.
+`ty` must be an object, so it must be an actual example of the data type rather
+than the name of the type. For example, for integer output, use `0::INTEGER`.
```
```suggestion
than the name of the type. For example, for integer output, use `0::INTEGER`.
Don't use `NULL::<TYPE>` as the example value. The function doesn't work correctly
with a `NULL` example.
```

charislam committed:
Apply suggestions from code review

Co-authored-by: Brian Rowe <brian@timescale.com>

charislam committed:
Fix <> that are messing with the parsing

WireBaron reviewed (approved):
There is some confusing behavior where the aggregate type returned by `freq_agg` and `topn_agg` are both `SpaceSavingAggregate` which is actually the input type for any of our frequency functions (like `topn`).  I'm not sure the easiest way to clarify this, but do feel that this is a good start.

charislam committed:
Merge branch 'latest' into charis/hyperfunctions-1-5

charislam committed:
Add frontmatter and update summaries

- Add and update hyperfunction frontmatter
- Update summary and page-index for counter aggs to include gauge_agg

charislam committed:
Change name of values to into_values

charislam committed:
Add into_values, min_frequency, max_frequency

WireBaron commented on api/counter_aggs.md:
```diff
@@ -1,18 +1,20 @@
-aggregation functions are used to continue accumulating data while ignoring any
-interruptions or resets. For more information about counter aggregation
-functions, see the [hyperfunctions documentation][hyperfunctions-counter-agg].
+# Counter and gauge aggregation
```
It might be worth noting that while the goal is to have the gauge agg support the entire counter agg API (with the possible exception of num_resets), the current experimental implementation only supports `delta`, `idelta_left`, and `idelta_right`.

WireBaron commented on api/into_values-freq_agg.md:
```diff
@@ -8,7 +8,7 @@ hyperfunction_subfamily: SpaceSavingAggregate
-# values()  <tag type="toolkit">Toolkit</tag><tag type="experimental">Experimental</tag>
+# into_values()  <tag type="toolkit">Toolkit</tag><tag type="experimental">Experimental</tag>
 This function returns the data accumulated in a 
 [frequency aggregate][freq_agg].  
```
This actually takes any SpaceSavingAggregate, so will work with `topn_agg` or `freq_agg`

WireBaron commented on api/gauge_agg.md:
```diff
@@ -1,3 +1,13 @@
+api_experimental: true
+hyperfunction_toolkit: true
+hyperfunction_family: 'counter aggregation'
+hyperfunction_subfamily: FrequencyAggregate
```
This subfamily seems wrong

WireBaron commented on api/counter_aggs.md:
```diff
@@ -1,18 +1,20 @@
-functions, see the [hyperfunctions documentation][hyperfunctions-counter-agg].
+# Counter and gauge aggregation
+This section contains functions related to counter and gauge aggregation.
+Counter aggregation functions are used to continue accumulating data while
+ignoring any interruptions or resets. Gauge aggregates are similar, but do not
+ignore resets. For more information about counter aggregation functions, see the
```
```suggestion
Counter aggregation functions are used to accumulate monotonically increasing data 
by treating any decrements as resets. Gauge aggregates are similar, but are used to
track data which can decrease as well as increase. For more information about counter
aggregation functions, see the
```

WireBaron commented on api/gauge_agg.md:
```diff
@@ -0,0 +1,84 @@
+api_category: hyperfunction
+api_experimental: true
+hyperfunction_toolkit: true
+hyperfunction_family: 'counter aggregation'
```
Would suggest something like 'metric aggregation'

WireBaron commented on api/min_frequency-max_frequency.md:
````diff
@@ -0,0 +1,63 @@
+) RETURNS DOUBLE PRECISION
+```
+
+`min_frequency` and `max_frequency` are accessors that operate on a [frequency
+aggregate][freq_agg]. Create a frequency aggregate over the original dataset,
````
```suggestion
`min_frequency` and `max_frequency` are accessors that operate on a [frequency
aggregate][freq_agg] or a top N aggregate. Create a frequency aggregate over the original dataset,
```

charislam commented on api/gauge_agg.md:
```diff
@@ -0,0 +1,84 @@
+|-|-|-|
+|`gauge_agg`|`GaugeSummary`|A `GaugeSummary` object that can be passed to accessor functions or other objects in the gauge aggregate API|
+
+The returned `GaugeSummary` can be used as an input to the same accessor
+functions as `CounterSummary`, except for `num_resets`.
```
```suggestion
<highlight type="important">
The returned `GaugeSummary` can be used as an input the accessor functions
`delta`, `idelta_left`, and `idelta_right`. When this feature is mature, it will support
all the same accessor functions as `CounterSummary`, with the exception of
`num_resets`.
</highlight>
```

charislam commented on api/counter_aggs.md:
```diff
@@ -26,7 +28,7 @@ additional hyperfunctions, you need to install the
 |||[`slope`](/hyperfunctions/counter_aggs/slope-counter/)|❌|✅|
 |||[`time_delta`](/hyperfunctions/counter_aggs/time_delta/)|❌|✅|
-|Counter aggregation|Counter aggregate mutators|[`with_bounds`](/hyperfunctions/counter_aggs/with_bounds/)|❌|✅|
+||Mutator|[`with_bounds`](/hyperfunctions/counter_aggs/with_bounds/)|❌|✅|
```
```suggestion
||Mutator|[`with_bounds`](/hyperfunctions/counter_aggs/with_bounds/)|❌|✅|

<highlight type="important">
All accessors can be used with `CounterSummary`. The accessors `delta`,
`idelta_left`, and `idelta_right` can be used with `GaugeSummary`.
`GaugeSummary` will eventually support all the accessors, with the exception
of `num_resets`.
</highlight>
```

charislam commented on api/gauge_agg.md:
```diff
@@ -0,0 +1,84 @@
+api_experimental: true
+hyperfunction_toolkit: true
+hyperfunction_family: 'counter aggregation'
+hyperfunction_subfamily: FrequencyAggregate
```
```suggestion
hyperfunction_subfamily: 'gauge aggregation'
```

charislam commented on api/into_values-freq_agg.md:
```diff
@@ -1,14 +1,14 @@
-# values()  <tag type="toolkit">Toolkit</tag><tag type="experimental">Experimental</tag>
+# into_values()  <tag type="toolkit">Toolkit</tag><tag type="experimental">Experimental</tag>
 This function returns the data accumulated in a 
 [frequency aggregate][freq_agg].  
```
```suggestion
[frequency aggregate][freq_agg] or [top N aggregate][topn_agg].  
```

charislam committed:
Apply suggestions from code review

Co-authored-by: Brian Rowe <brian@timescale.com>

charislam committed:
Add topn_agg to values and min/max

charislam committed:
Rename counter agg family to metric agg

Rename the 'counter aggregation ' family to be more generic
to accommodate introduction of gauge_agg

WireBaron commented on api/min_frequency-max_frequency.md:
````diff
@@ -0,0 +1,65 @@
+```
+
+`min_frequency` and `max_frequency` are accessors that operate on a
+[frequency aggregate][freq_agg] or a top N aggregate[topn)agg]. Create a
````
```suggestion
[frequency aggregate][freq_agg] or a [top N aggregate][topn_agg]. Create a
```

charislam committed:
Fix typos

Co-authored-by: Brian Rowe <brian@timescale.com>

charislam committed:
Merge branch 'latest' into charis/hyperfunctions-1-5

Loquacity committed:
Merge branch 'latest' into charis/hyperfunctions-1-5

charislam committed:
Merge branch 'latest' into charis/hyperfunctions-1-5
